File: tools/@penn_crop/penn_crop.m
classdef penn_crop
    properties (Access = private)
        opt
        split
        dir
        ind2sub
        visible
        part
        seqId
        nFrame
        nPhase
        seqType
        seqLength
        inputRes
        outputRes
    end
    methods
        % constructor
        function obj = penn_crop(opt, split)
            obj.opt = opt;
            obj.split = split;
            obj.dir = fullfile(opt.data, 'frames');
            assert(exist(obj.dir,'dir') == 7, ['directory does not exist: ' obj.dir]);
            % load annotation
            annot_file = fullfile(opt.data, [split '.h5']);
            obj.ind2sub = permute(hdf5read(annot_file,'ind2sub'),[2 1]);
            obj.visible = permute(hdf5read(annot_file,'visible'),[2 1]);
            obj.part = permute(hdf5read(annot_file,'part'),[3 2 1]);
            % preprocess annotation
            [obj.seqId, obj.nFrame] = preproAnno(obj);
            % get phase number and LSTM sequence length
            obj.nPhase = opt.nPhase;
            obj.seqType = opt.seqType;
            obj.seqLength = opt.seqLength;
            % get input and output resolution
            obj.inputRes = opt.inputRes;
            obj.outputRes = opt.outputRes;
        end
        
        % get seq id and num of frames
        [seqId, nFrame] = preproAnno(obj);
        
        % get phase sequence in global index (ind2sub)
        [ind, has_flow] = getSeq(obj, i);
        
        % get image path
        pa = imgpath(obj, idx);
        
        % load image
        img = loadImage(obj, idx);
        
        % get center and scale
        [center, scale] = getCenterScale(obj, img);
        
        % get dataset size
        out = size(obj);
        
        [input, phaseSeq, center, scale] = get(obj, idx);
        
        [sid, fid] = getSeqFrId(obj, idx);
        
        % get sampled indices; for prediction and visualization
        sidx = getSampledIdx(obj);
    end
end

File: caffe-feat/lib/datasets/penn-crop.lua
require 'hdf5'
require 'image'

local t = require 'lib/datasets/transforms'

local M = {}
local PennCropDataset = torch.class('caffe-feat.PennCropDataset', M)

function PennCropDataset:__init(opt, split)
  self.split = split
  self.dir = paths.concat(opt.data, 'frames')
  assert(paths.dirp(self.dir), 'directory does not exist: ' .. self.dir)
  -- Load annotation
  annot_file = paths.concat(opt.data, split .. '.h5')
  self.ind2sub = hdf5.open(annot_file,'r'):read('ind2sub'):all()
  -- Get resize and crop size
  self.resizeRes = opt.resizeRes
  self.cropRes = opt.cropRes
  -- Get mean
  self.mean = opt.mean
end

-- Get image path
function PennCropDataset:_imgpath(idx)
  return string.format('%04d/%06d.jpg',self.ind2sub[idx][1],self.ind2sub[idx][2])
end

-- Load image
function PennCropDataset:_loadImage(idx)
  return image.load(paths.concat(self.dir,self:_imgpath(idx)), 3, 'byte')
end

function PennCropDataset:_preprocess(im, train, flip)
  local input = im
  input = image.scale(input, self.resizeRes, self.resizeRes)
  input = input:float()
  for i = 1, 3 do
    input[i]:csub(self.mean[i])
  end
  input = input:index(1,torch.LongTensor({3,2,1}))
  input = t.TenCrop(self.cropRes)(input)
  return input
end

-- Get dataset size
function PennCropDataset:size()
  return self.ind2sub:size(1)
end

function PennCropDataset:get(idx)
  -- Load image
  local im = self:_loadImage(idx)

  -- Preprocess input
  local input = self:_preprocess(im)

  return {
    input = input,
  }
end

function PennCropDataset:getSeqFrId(idx)
  return self.ind2sub[idx][1], self.ind2sub[idx][2]
end

return M.PennCropDataset